src/r_exasol/ssl/certificate.cpp

```diff
@@ -0,0 +1,169 @@
+
+        class CertExtensionBuilder {
+        public:
+            explicit CertExtensionBuilder(X509 *cert) : mCert(cert) {}
+
+            CertExtensionBuilder& operator << (const CertExtension && certExtension);
+        private:
+            X509 *mCert;
+        };
+
+        /* Add extension using V3 code: we can set the config file as NULL
+         * because we wont reference any other sections.
+         */
+
+        CertExtensionBuilder &CertExtensionBuilder::operator<<(const CertExtension &&certExtension) {
+            X509_EXTENSION *ex;
+            X509V3_CTX ctx;
+            /* This sets the 'context' of the extensions. */
+            /* No configuration database */
+            X509V3_set_ctx_nodb(&ctx);
+            /* Issuer and subject certs: both the target since it is self signed,
+             * no request and no CRL
+             */
+            X509V3_set_ctx(&ctx, mCert, mCert, nullptr, nullptr, 0);
+            ex = X509V3_EXT_conf_nid(nullptr, &ctx, certExtension.mNid, certExtension.mValue);
+            if (!ex) {
+                throw CertificateException("Cannot add extension to certificate.");
+            }
+            X509_add_ext(mCert,ex,-1);
+            X509_EXTENSION_free(ex);
+            return *this;
+        }
+
+        struct CertName {
+            CertName(const char * field, int type, const char * bytes)
+            : mField(field)
+            , mType(type)
+            , mBytes(bytes) {}
+            const char * mField;
+            int mType;
+            const char * mBytes;
+        };
+
+        class CertNameBuilder  {
+        public:
+            CertNameBuilder(X509_NAME *name) : mName(name) {}
+
+            CertNameBuilder & operator <<(const CertName && certName);
+
+        private:
+            X509_NAME *mName;
+        };
+
+        CertNameBuilder &CertNameBuilder::operator<<(const CertName &&certName) {
+            const unsigned char* bytes = reinterpret_cast<const unsigned char *>(certName.mBytes);
+            int ret = X509_NAME_add_entry_by_txt(mName, certName.mField,
+                                                 certName.mType, bytes, -1, -1, 0);
+            if (1 != ret) {
+                throw CertificateException(std::string("cannot set name of certificate for field:") + certName.mField);
+            }
+            return *this;
+        }
+    }
+}
+
+void exa::ssl::Certificate::mkcert(int bits, int serial, int days) {
+    X509 *x;
+    EVP_PKEY *pk;
+    RSA *rsa = nullptr;
+    X509_NAME *name=nullptr;
+
+    pk=EVP_PKEY_new();
+    x=X509_new();
+
+    rsa = RSA_new();
+
+    BIGNUM* bn = 0;
+    bn = BN_new();
+    BN_set_word(bn, RSA_F4);
+    int ret = RSA_generate_key_ex(rsa, bits, bn, nullptr);
+    BN_free(bn);
+    if (!ret) {
+        throw CertificateException("Error generating RSA key.");
+    }
+    if (!EVP_PKEY_assign_RSA(pk,rsa))
+    {
+        throw CertificateException("Error assigning RSA key.");
+    }
+    rsa=nullptr;
```

Comment:
Is this enough to remove the RSA key from memory?

Reply:
Actually I was questioning myself the same thing when I saw the demo code.
But the documentation says:

> EVP_PKEY_assign_RSA(), EVP_PKEY_assign_DSA(), EVP_PKEY_assign_DH(), EVP_PKEY_assign_EC_KEY(), EVP_PKEY_assign_POLY1305() and EVP_PKEY_assign_SIPHASH() also set the referenced key to key however these use the supplied key internally and so key will be freed when the parent pkey is freed.

So believe it is correct.